### ISSUE TITLE
Align explorer sorting with platform sorting

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#27759]
It looks like our file sorting in the explorer does not match platform beahviour in some cases.

**Windows:**
* a file `foo.ts` is sorted before `foo_test.ts` but we sort it the other way around

**Linux:**
* a file `foo.ts` is sorted before `foo_test.ts` but we sort it the other way around
* a lowercase file seems to be sorted before an upper case file but we seem to mix the sorting independent of the casing (e.g. folders `[out, outb, outd, Outa, Outc]` are showing up as `[out, OutA, outb, Outc, outd]`

**macOS:**
* seems to be OK

We use a JavaScript [`Collator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Collator) for the comparing [here](https://github.com/Microsoft/vscode/blob/master/src/vs/base/common/comparers.ts#L19). 

Unfortunately I am not able to tweak the Collator options to bring me the desired result...

**Comment by comerc** [comment]
I want sort order like GitHub repo. Please!

**Comment by NickWest-appuri** [comment]
![screen shot 2017-08-29 at 1 20 55 pm](https://user-images.githubusercontent.com/30420175/29842203-ff6cf16a-8cbc-11e7-939b-528166217e8b.png)
I have some YML files whose names are based on GUIDs.  They aren't even close to being alpha sorted. 

vscode 1.15.1, macOS Sierra 10.12.6

**Comment by dlech** [comment]
> Linux: a lowercase file seems to be sorted before an upper case file

Umm... I think it is the other way around, upper case first. Basically, just sorting using the ASCII value of each character.

**Comment by wildcart** [comment]
@dlech  on Linux, at least, it depends on the chosen 'locale' and the environment variable `LC_COLLATE` can be used to influence this behaviour (this influences for example the `ls` command, ).

If you consider to implement / support platform specific behaviour you may want to consider evaluating the locale setting on Linux, specifically LC_COLLATE (if set, otherwise fallback to the set local).

```
user@host -- ~/tmp/casetest $ LC_COLLATE='en_GB.UTF-8' ls -1
2
5
a
A
Aa
aB
AZ
C
user@host -- ~/tmp/casetest $ LC_COLLATE='en_EN.UTF-8' ls -1
2
5
A
AZ
Aa
C
a
aB
user@host -- ~/tmp/casetest $ LC_COLLATE='C' ls -1
2
5
A
AZ
Aa
C
a
aB
```

**Comment by leidegre** [comment]
It's called shortlex, or lexiographic sort order. The only thing you need to tweak is the string length. If A is shorter than B then A is smaller than B. This is not going to be covered by any collation. An alternative to this is to introduce padding (padding of the sort so that the comparison is less) but I don't think it's reasonably to do that due to the extra garbage generated.

**Comment by egonelbre** [comment]
Shortlex orders primarily by length, the code presented implements a different ordering.

```
aggregate.go
aggregate_registry.go
event.go
event_registry.go
README.md
rehydration.go
```

as

```
event.go
README.md
aggregate.go
rehydration.go
event_registry.go
aggregate_registry.go
```

Windows Explorer uses [Natural Sort](https://en.wikipedia.org/wiki/Natural_sort_order), because length is not sufficient for good order, as an example:

```
action_5_example.txt
action_10_ex.txt
```

**Comment by leidegre** [comment]
@egonelbre then I misunderstood the meaning of shortlex, I should have just said lexicographic. but the code does what it is supposed to do. that is, first sort up to X characters, then use the length as a discriminator.

**Comment by leidegre** [comment]
Why complicate this though. Lexicographic (not shortlex as I incorrect first called) it easy to implement and understand. This is not going to get done if we insist on extra work to align with something which is highly Windows Explorer specific. That should not be the high water mark here.

**Comment by egonelbre** [comment]
It's not really Explorer specific, you can read more about it in https://blog.codinghorror.com/sorting-for-humans-natural-sort-order/. I mentioned it because your other issue showed that as an example.

The reason you don't want to use lexicographic first is due due to the last example.

```
action_5_example.txt
action_10_ex.txt
```

Sort lexicographically is:

```
action_10_ex.txt
action_5_example.txt
```

**Comment by leidegre** [comment]
@egonelbre as a programmer, I don't care. But as a user of Windows Explorer, I could see why adhering to the natural sort order would seem more natural.

**Comment by fenchu** [comment]
I tried making this: https://github.com/microsoft/vscode/issues/75415 but was closed. There is currently no option to make the explorer sort native.

and no-one writes 5, 10 we write 05,10, this is a fundamental fact.

**Comment by AnyhowStep** [comment]
vs code sorts like this,
```
log-2
log-10
```

I'd rather it sort like this,
```
log-10
log-2
```

It is **very jarring** for me when vs code does not sort lexicographically.

Almost every other tool I use sorts lexicographically. When vs code tries to be different, it just confuses me for a short moment. Repeatedly.

**Comment by leilapearson** [comment]
@bpasero I'm hoping you can advise and save me some time if this doesn't make sense or is unlikely to be accepted as a pull request.

I'm considering creating a pull request that would add a new setting - `explorer.sortCaseSensitive` with a default value of `false`.  

I considered adding one or more options to the existing `explorer.sortOrder` setting, but case sensitivity seems to be orthogonal to those options - and (nearly) doubling the number of options to add case sensitive versions doesn't seem like the best idea:

```typescript
'explorer.sortOrder': {
  'type': 'string',
  'enum': [SortOrder.Default, SortOrder.Mixed, SortOrder.FilesFirst, SortOrder.Type, SortOrder.Modified],
  'default': SortOrder.Default,
  'enumDescriptions': [
    nls.localize('sortOrder.default', 'Files and folders are sorted by their names, in alphabetical order. Folders are displayed before files.'),
    nls.localize('sortOrder.mixed', 'Files and folders are sorted by their names, in alphabetical order. Files are interwoven with folders.'),
    nls.localize('sortOrder.filesFirst', 'Files and folders are sorted by their names, in alphabetical order. Files are displayed before folders.'),
    nls.localize('sortOrder.type', 'Files and folders are sorted by their extensions, in alphabetical order. Folders are displayed before files.'),
    nls.localize('sortOrder.modified', 'Files and folders are sorted by last modified date, in descending order. Folders are displayed before files.')
  ],
  'description': nls.localize('sortOrder', "Controls sorting order of files and folders in the explorer.")
},
```

This change would only partially address this open issue, since it:

- only addresses case sensitivity
- doesn't default to align with platform case sensitivity

I think it would probably satisfy a lot of people though, and TBH I'm not sure that aligning with platform case sensitivity is the best option. For example, I noticed a number of comments on this and related issues were asking to align the file sort order with github, which always does a case sensitive sort. 

Anyway, by providing a setting, people can choose, and by defaulting to the current behavior nobody will be affected by the change unless they want to be.

Also, if the future default behavior is changed, this setting will still be useful for people who want to override that default behavior.

What do you think? Should I go ahead? 

If yes, is this the right issue to reference in the PR or should I create a separate issue that just links to this one?

Thanks in advance!

**Comment by isidorn** [comment]
@leilapearson thanks for the offer.
Ideally we would just align explorer sorting with platform sorting without any option. @bpasero already provided a code pointer where he is doing the comparing.

If that is not possible only then we can look into adding more settings.

An alternative is to look to open this up to extensions and then extensions could control this and satisfy the 20 different sorting styles that users want.

**Comment by leilapearson** [comment]
@isidorn thanks for the reply. That's why I asked before doing anything other than taking a look at the code.

Opening this to extensions is an interesting option. At the same time, I would still think that offering control over whether the sort is case sensitive or not should be a core option and not require an extension.

It isn't easy on some platforms to adjust the sort order - and having to figure out how to get your whole platform to sort case sensitive in order for VS Code to sort case sensitive seems a bit awkward? Especially if you primarily develop on one platform and only spend a bit of time developing on other platforms.  

Also, I find that programming on a platform is a different context than using a platform for office work. Having different sort orders apply to the different contexts often makes sense.  For example, I tend to sort things by most recently modified when I'm working on documents and the like - so this is my default in file explorer and google docs - but I don't want my code sorted by modification date and I'm happy with how things are sorted in my terminal - but unfortunately not so happy with how they are sorted in VS Code.

I do agree that too many settings can be a bad thing, but I'm curious if you agree or not that a setting to control case sensitivity would make sense regardless?

**Comment by leilapearson** [comment]
P.S. An example of how hard it can be to change the sort (collate) order on a platform is OSX - which doesn't expose any nice way to do that it seems:

https://apple.stackexchange.com/questions/34054/case-insensitive-ls-sorting-in-mac-osx

**Comment by isidorn** [comment]
Ok, makes sense. I would be open to a lean and nice PR that controls if sorting is case sensitive or not. 
Thanks

**Comment by leilapearson** [comment]
Thanks @isidorn. Before I go ahead, a bit of extra context and one more question...

There are actually a total of 4 options for defining what "alphabetically" means. Per the [ECMA 402 standard](https://tc39.es/ecma402/#sec-intl.collator.prototype.constructor):

> The sensitivity of collator is interpreted as follows:
>
> **base**: Only strings that differ in base letters compare as unequal. Examples: a ≠ b, a = á, a = A.
> 
> **accent**: Only strings that differ in base letters or accents and other diacritic marks compare as unequal. Examples: a ≠ b, a ≠ á, a = A.
> 
> **case**: Only strings that differ in base letters or case compare as unequal. Examples: a ≠ b, a = á, a ≠ A.
> 
> **variant**: Strings that differ in base letters, accents and other diacritic marks, or case compare as unequal. Other differences may also be taken into consideration. Examples: a ≠ b, a ≠ á, a ≠ A.
> 
> **NOTE**  In some languages, certain letters with diacritic marks are considered base letters. For example, in Swedish, "ö" is a base letter that's different from "o".
> 

Instead of just exposing case sensitivity as a `true` or `false` option, I'm thinking it would be best to allow any of the 4 options. `base` would be the default value since that's what's hardcoded into the current code.

Any objection to offering all 4 options?

**Comment by isidorn** [comment]
@leilapearson this makes sense. However to simplify this a bit I suggest the following:
* we introduce a `explorer.sortCaseSensitive` setting with string values "on" and "off"
* if users request for this other options we can add them later (since we sue a `string` setting this should not be a problem)

The reason why I prefer this solution is simplicity and I think it covers the 99% use case.
Let me know what you think.

**Comment by leilapearson** [comment]
Sounds good. Thanks @isidorn .

**Comment by leilapearson** [comment]
Well @isidorn, that was a bit trickier than expected, but I think I have something that is almost ready to submit.  I'll take one last look tomorrow to make sure I haven't missed anything. 

Unfortunately the solution I was originally picturing didn't work. It turns out that grouping by case is a very different thing than comparing by case! :-) 

There were also some special cases that I needed to adjust for - including the one that @leidegre pointed out. I can describe them next to the relevant code when I submit the PR.

Since the solution was different than I was imagining, the new setting is different than we discussed, but I think it's still simple to understand and use. Let me know if you have any concerns or comments.

Here's what it looks like now:
```ts
'explorer.sortOption': {
  'type': 'string',
  'enum': [SortOption.Numeric, SortOption.Upper, SortOption.Lower, SortOption.Mixed],
  'default': SortOption.Numeric,
  'enumDescriptions': [
    nls.localize('sortOption.numeric', 'Mixes uppercase and lowercase names together. Numbers are sorted numerically, not alphabetically.'),
    nls.localize('sortOption.upper', 'Groups uppercase names before lowercase names. Numbers are sorted alphabetically.'),
    nls.localize('sortOption.lower', 'Groups lowercase names before uppercase names. Numbers are sorted alphabetically.'),
    nls.localize('sortOption.mixed', 'Mixes uppercase and lowercase names together. Numbers are sorted alphabetically.')
  ],
  'description': nls.localize('SortOption', "Further specifies the file and directory sort order.")
}
```

I'm very happy to say that with this new setting and some small tweaks in the code I believe that the whole problem might be solved - or at least solved enough to satisfy most people. 

Namely:

- String comparisons use the platform locale
- `aggregate.go` and `aggregate_repo.go` sort as expected
- filenames that start with a dot (hidden files) sort as expected
- between the existing `SortOrder` and the new `SortOption` setting, users should be able to emulate the most popular filename grouping options - including the one that github uses.

Whew!

**Comment by leilapearson** [comment]
Actually I just realized one more option would be good to add - namely a simple unicode sort. 

It seems that the terminal on a Mac uses a simple unicode sort. The file explorer on a Mac uses a localized sort though.  

I think a lot of people might want to match their sort order to their terminal, and the other options don't give you that. 

It would be trivial to add a `SortOption.Unicode` to the list.

**Comment by isidorn** [comment]
Sound good. Once you submit a PR feel free to ping me @isidorn on it and we can continue the discussion there. Thanks

**Comment by leilapearson** [comment]
Perfect @isidorn . Thanks!

**Comment by Sytten** [comment]
Any progress?

**Comment by leilapearson** [comment]
@Sytten some sort order edge cases were addressed in #97200 and that change is available in vscode 1.46.0. See the PR for a detailed description.

I also have an open PR #97272 - old now and sure to need an update - to add some additional lexicographic options to allow sorting in unicode order, locale order with uppercase first, or locale order with lowercase first.

Which specific functionality were you hoping to see addressed?

**Comment by Sytten** [comment]
On MacOS the files are sorted case insensitive and I didnt find a way to sort them case sensitive (without affecting the order files/folders).

**Comment by leilapearson** [comment]
PR #97272 adds the option to group files and folders by case, but that PR was submitted at a time when the reviewers weren't available and is out of date now. I'll take a look at resurrecting it.

**Comment by leilapearson** [comment]
Just wanted to provide a couple of quick updates for anyone watching this issue.

1. PR #104528 was recently merged. This PR changes how `aggregate.go` and `aggregate_repo.go` are sorted. See Issue #99955 if you want more details.
2. PR #97272 - which adds settings to group names by case and to sort in unicode order - has been updated, but will be left on hold for now.

**Comment by Papooch** [comment]
These discussions are taking years and new sorting order proposals are coming up. I think most sort options are so niche, that I think it would be better for an extension to provide them instead of putting them in the core. Will there ever be an option to provide a custom sorting algorithm, or a way for an extension to add one?

**Comment by leilapearson** [comment]
Just a quick mention that PR #97272 was merged and released a while back. 

This may resolve your issue @Sytten

**Comment by Sytten** [comment]
Thanks @leilapearson I will try it 👍

**Comment by ALIENQuake** [comment]
Having aligned explorer sorting with platform sorting is not 'niche', 

Another example:
![image](https://user-images.githubusercontent.com/365255/129192574-f163b4b0-2fd1-456f-8980-e6438ed4738a.png)

Please fix this!

**Comment by Mushr0000m** [comment]
![Capture d’écran 2022-02-14 à 16 02 35](https://user-images.githubusercontent.com/5532323/153888886-c433cdd5-ae34-4935-90c3-3d9321d2786f.png)

Same problem here shorter names are not first, `prospect-edit.scss` should be before `prospect-edit-user.scss`.

**Comment by leilapearson** [comment]
I agree it would be nice to be able to match the Windows file explorer order when desired (and probably by default when on Windows). It's complicated though and would require a function that can translate from unicode values to Windows Explorer sort order values - taking locale into account if necessary. 

I wasn't able to find such a function, and implementing one without actually having the code used in Windows Explorer seems like it would require a lot of time-consuming experimentation with different characters and locales. That said, vscode is made by Microsoft, so someone on the team may be able to use their contact to get access to the relevant Windows Explorer function to make this easier :-)

For my own possibly future reference - and maybe your interest - I've copied in a table showing one person's investigation

The table comes from:

https://superuser.com/questions/681322/windows-explorer-sorting-order-for-special-characters

```
Characters Allowed in File Names (sorted in File Explorer collating order)

Unicode
Character  Hex Value     Description
---------  ------------  ----------------------------------------
!          0021          exclamation mark
#          0023          number sign
$          0024          dollar sign
%          0025          percent sign
&          0026          ampersand
(          0028          left parenthesis
)          0029          right parenthesis
,          002C          comma
.          002E          full stop, period
'          0027          apostrophe
-          002D          hyphen, minus
;          003B          semicolon
@          0040          commercial at sign
[          005B          left square bracket
]          005D          right square bracket
^          005E          circumflex accent
_          005F          low line, underscore
`          0060          grave accent
{          007B          left curly bracket
}          007D          right curly bracket
~          007E          tilde
+          002B          plus sign
=          003D          equal sign
0-9        0030 – 0039   digit zero through digit nine
A-z        0041 – 005A,  capital letter A through Z
           0061 – 007A   small letter a through z

Note: File Explorer does not differentiate between capital and small letters in file names, i.e. ‘A’ and ‘a’ are considered to be the same character.
```

You can see in the above that Windows order is `. _ -` while unicode order `- . _`. I also know that my locale order is ` _ - .` - so different yet again. I'm not sure if the actual order of punctuation characters in Windows Explorer also varies by locale or not.

Anyway, I'm open to helping out with this if someone can find me a function for reference :-)

**Comment by rodneyazev** [comment]
I was just thinking that @leilapearson

Sorting these files in alphabetical order would help us keep our codes organized,  just like Windows Explorer, Spring Boot and so many IDE/Editors.

Not only could they get this functionality by taking some of the code contained in Windows Manager, but they could see the source code for so many open source tools available. ls itself can do this. Even via javascript it is easy to do this.

I don't know if they are undertanding, but the main problem with not being organized is maintaining or supporting thousands of kubernetes environments, code files, etc. It would be very complicated to provide adequate support when projects become complex (huge).

If they have managed to make a "supposed organization", they could make a better one, right?

**Comment by omBratteng** [comment]
I've noticed, that with the default `explorer.sortOrder` settings, files starting with `aa` comes after `az`.
Example:
```console
touch {a,aa,ab,az}
```
<img width="199" alt="image" src="https://user-images.githubusercontent.com/1681525/162186507-3a411a49-016e-4559-966d-dab14f9ea892.png">

Using macOS with Norwegian locale. But in finder, `aa` comes before `ab`, as expected.
<img width="131" alt="image" src="https://user-images.githubusercontent.com/1681525/162186659-7c966df7-02b3-44d0-b32c-018fc7c79224.png">

**Comment by omBratteng** [comment]
That’s correct, aa is pronounced as å, wasn’t till 1917 we converted from aa to å. Denmark in 1948.
I bet if you did the same with Danish, you would get same results as Norwegian.

Technically, it is correct. Aa is Å, and should be sorted with Å.
> Correct [alphabetization](https://en.m.wikipedia.org/wiki/Danish_and_Norwegian_alphabet) in Danish and Norwegian places Å as the last letter in the alphabet, the sequence being [Æ](https://en.m.wikipedia.org/wiki/%C3%86), [Ø](https://en.m.wikipedia.org/wiki/%C3%98), Å. This is also true for the alternative spelling "Aa". Unless manually corrected, sorting algorithms of programs localised for Danish or Norwegian will place e.g., Aaron after Zorro.

I just encountered this when I was directory named [aad-pod-identity](https://github.com/Azure/aad-pod-identity), and couldn’t understand why it ended up under `wordpress` in the list.

Perhaps a setting in VS Code that lets me override which locale to sort with. Or just using the `"explorer.sortOrderLexicographicOptions": "unicode"` setting, which put them in the “correct” order as intended.

**Comment by George-Ogden** [comment]
One simple solution in the short term would be a `"unicode-case-insensitive"` ordering that puts `file.a` before `file_a` but keeps `filea` before `fileB`.

**Comment by php4fan** [comment]
> [Linux] a lowercase file seems to be sorted before an upper case file but we seem to mix the sorting independent of the casing

No, the current default is the right one  (sort regardless of case: that's how almost every application does it, except those that haven't evolved since the 80s)  and there's already a setting that allows you to change between lowercase first, uppercase first or "default" which is mixed. That setting BTW also has a fourth option "unicode", which is stupid, because whether or not the sorting uses unicode (which seems like it should be the default) should be independent from how to sort case.

**Comment by mvaneerde** [comment]
**Concrete use case: folders named with GUIDs**

A folder full of subfolders whose names are GUIDs (session IDs, correlation IDs, etc.) gets mis-sorted by natural sort because the leading hex digits are interpreted as numbers.

Lexicographic order (expected):

```
03806eca-a8b9-4e02-8d15-259406aad047
109222f2-7a36-4f76-b250-20e499c0b2e5
60af07ff-ec2b-4a48-960c-b803efff5881
707a3abb-75a8-43e1-b00a-e508d94137a6
a639c873-20c3-4931-a61b-2b0abd860d7d
```

Natural sort order (what VS Code shows):

```
60af07ff-ec2b-4a48-960c-b803efff5881
707a3abb-75a8-43e1-b00a-e508d94137a6
03806eca-a8b9-4e02-8d15-259406aad047
109222f2-7a36-4f76-b250-20e499c0b2e5
a639c873-20c3-4931-a61b-2b0abd860d7d
```

`03806eca` sorts *after* `707a3abb` because natural sort reads 3806 > 707. The displayed order is neither lexicographic, chronological, nor meaningful in any way. A strict lexicographic / ASCIIbetical sort option would fix this.
